subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy

```diff
@@ -1022,6 +1022,98 @@ class ArtifactTransformIntegrationTest extends AbstractHttpDependencyResolutionT
         outputContains("ids: [producer2.jar (producer2.jar)]")
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/22735")
+    def "transform selection prioritizes shorter transforms over attribute schema matching"() {
+        buildFile << """
+            project(':lib') {
+                task jar(type: Jar) {
+                    destinationDirectory = buildDir
+                    archiveFileName = 'lib.jar'
+                }
+
+                configurations.compile.outgoing.variants{
+                    primary {
+                        attributes {
+                            attribute(artifactType, "jar")
+                            attribute(extraAttribute, "preferred")
+                        }
+                        artifact jar
+                    }
+                    secondary {
+                        attributes {
+                            attribute(artifactType, "intermediate")
+                        }
+                        artifact jar
+                    }
+                }
+            }
+
+            // Provides a default value of `preferred` if a given attribute is not requested.
+            abstract class DefaultingDisambiguationRule implements AttributeDisambiguationRule<String>, org.gradle.api.internal.ReusableAction {
```

Comment:
Good catch. We would not. Before these changes, the longer path was preferred due to the disambiguation rule. Now, the longer path is not even found when searching. So, this WOULD fail in master, but now with these changes the disambiguation rule is not even necessary. 

I'm inclined to leave this in, just as a regression test against the previous functionality. 

Thinking out loud, the only way we could make this test more relevant to current functionality is to have two transforms of the same length being disambiguated by the disambiguation rule. Though, in that case, the one with the preferred attribute WOULD be selected as opposed to the current test, where the one with the preferred attribute is not selected.

Reply:
Agree, let's leave this here.

Here's the test failing with master in the way you described: https://ge.gradle.org/s/wk54kjplibz44/tests/:dependency-management:embeddedIntegTest/org.gradle.integtests.resolve.transform.ArtifactTransformIntegrationTest/transform%20selection%20prioritizes%20shorter%20transforms%20over%20attribute%20schema%20matching?top-execution=1